### DISCUSSION TITLE
android 10 branch forking

### DISCUSSION THREAD

**Discussion by simbadMarino** [termux/termux-app#2441] [Q&A] [answered]
Hello termux devs and community, I finally decided to move from updated flavor branch to android10 as per @agnostic-apollo recommendation but I’m having some issues when running the TermuxActivity. Not sure if I need to modify the $PREFIX at `android/pkg.sh `. It’s not executing the login as usual and everything looks empty. Seems like symlinks are not available as well. I hope anyone here can give me a light on what could be the root cause of the issue. I’m adding a screenshot and execution log file
![1A5E2FD4-697F-4C66-AFFC-11BE3BE6DF9D](https://user-images.githubusercontent.com/11146636/144337869-df80409d-d6aa-4f15-9366-38fc908f6e31.jpeg)

[android10 log.txt](https://github.com/termux/termux-app/files/7638074/android10.log.txt)

**Comment by agnostic-apollo** [comment, marked as answer]
Didn't you get the `updated-flavor` branch working after the patches?

**Reply by simbadMarino** [reply]
Yes , however I was having issues when building android bundles, with the bundles the app is always calling the old bootstrap setup, so I decided to make a “cleaner” fork with this android10 branch :)

**Reply by agnostic-apollo** [reply]
I told you this before.

> In `android-10` branch, the `TermuxActivity` does not call `TermuxInstaller` to set up the bootstrap, but instead `TermuxService` calls [`TermuxPackageInstaller.setupAllInstalledPackages(this)`](https://github.com/termux/termux-app/blob/e783f40670ddfedab46e010e661e6ce67d59be63/app/src/main/java/com/termux/app/TermuxService.java#L245) which then installs `com.termux` package itself to setup bootstrap. 

>In `updated-flavor` branch, the TermuxActivity is [still calling `TermuxInstaller`](https://github.com/termux/termux-app/blob/60f07718880bdd8d241f3c13126343b0b406c818/app/src/main/java/com/termux/app/TermuxActivity.java#L258) to set up the bootstrap in the old way and `TermuxService` [starts `TermuxPackageInstallerService`](https://github.com/termux/termux-app/blob/60f07718880bdd8d241f3c13126343b0b406c818/app/src/main/java/com/termux/app/TermuxService.java#L119) which then also [installs all packages](https://github.com/termux/termux-app/blob/updated-flavor/app/src/updated/java/com/termux/app/TermuxPackageInstallerService.java#L47), including `com.termux`.

>That is likely the cause of the problem, due to conflicting bootstrap installation. The bootstrap zip of both `--android10` and default is the same and has `SYMLINKS.txt`, so `TermuxInstaller` would succeed. The `build.gradle` adds the `libfiles.so` to the zip in `android-10` and `updated-flavor` branches so `TermuxPackageInstaller` "should" succeed too, check logs for errors. Not sure what the end result is for dual installation, but seeing it’s not working for you, my guess is that `TermuxInstaller` runs first. You can try not calling it in `TermuxActivity` instead, you can compare changes with `android-10` branch.

Moreover, its highly advisable not to release such old versions of `termux-app` to the public, wait till I do some basic merging of branches so you and others can fork `master` itself, I'll try doing it for next `v0.118` release if time and resources allows it.

**Reply by simbadMarino** [reply]
Agree! I was hesitant to use the android10 fork for the same reason(being older than The updated-flavor) . Regarding the suggestion you made I tried to deactivate it but encounter other issues and preferred to do the android 10 fork thing 😅, if you are merging to master sooner or later then I believe it’s the best for me to wait for it to keep compliant with termux plans.

**Reply by agnostic-apollo** [reply]
What other issues are you encountering? For current testing purposes, merging would be close to `updated-flavor` branch than `android-10`, so best use it so that moving to `master` later is easier.

Yes, it is planned, but priorities change, specially if its not highly requested/important. I also need to think about support for other "flavours" like multi-user, non-termux based on android `/system/bin` only. So design needs to be thought through since its going to be complex and so time consuming.

**Reply by simbadMarino** [reply]
Let me revert changes and will let you know exactly the issues here

**Reply by simbadMarino** [reply]
hello again @agnostic-apollo , I reverted the changes and using updated-flavor again. I removed the TermuxInstall part as you requested as follows: 
![image](https://user-images.githubusercontent.com/11146636/144604028-6d3195c3-b7f7-44f6-a0a4-aade0648f090.png)

I tested it with the Default APK run configuration and works like a charm with the following 'run' log:
[android_apk_build.log](https://github.com/termux/termux-app/files/7649644/android_apk_build.log)

However if test with the 'APK from bundle' run option I get the following after opening TermuxActivity:
![image](https://user-images.githubusercontent.com/11146636/144604637-1c6b8d58-c128-4abb-82cc-bccaafc46ded.png)

and this log:
[android_bundle_build.log](https://github.com/termux/termux-app/files/7649653/android_bundle_build.log)

I'm becoming clueless on what's causing this different behavor (Default APK build vs Android Bundle)... thanks!

**Reply by agnostic-apollo** [reply]
I am pretty sure you have not fixed the `com.termux` hardcoded values [here](https://github.com/termux/termux-app/blob/60f07718880bdd8d241f3c13126343b0b406c818/app/src/main/java/com/termux/app/TermuxService.java#L119) and [here](https://github.com/termux/termux-app/blob/60f07718880bdd8d241f3c13126343b0b406c818/app/src/updated/java/com/termux/app/TermuxPackageInstaller.java#L129) since no [`Installing: `](https://github.com/termux/termux-app/blob/60f07718880bdd8d241f3c13126343b0b406c818/app/src/updated/java/com/termux/app/TermuxPackageInstaller.java#L55) log entries exist. Also go in termux settings, long hold terminal -> `More` -> `Settings` and enable `Verbose` logging and then restart termux, it should show more details for `TermuxSession` but not package installation. Also run from failsafe session `ls -a /data/data/com.justshare/files/usr/bin` and see if symlinks exist.

Moreover, with current broken design, all packages will be reinstalled on every startup. which would result in slow startup, specially if lot of packages exist. So such stuff needs to handled somehow as well before merging, likely adding an option in settings to install all packages instead of doing it every time and only installing `com.termux` if `$PREFIX` does not exist and installing other packages only when their install intent is received. If users install other packages before `com.termux`, then they will have to use the settings option to install them manually or I guess can be handled during bootstrap at install time by installing all packages.

**Reply by simbadMarino** [reply]
I just double checked and confirmed  I have the correct packages, the issue seems to be the symlinks are non existant.( I ran the `ls -a /data/data/com.justshare/files/usr/bin` and symlinks DO NOT exist.

When filesMappingFile.exists() is evalueted I'm getting a 0 in the following code:

static void installPackage(Context context, ApplicationInfo info) throws Exception {
        File filesMappingFile = new File(info.nativeLibraryDir, "libfiles.so");
        Log.e("termux",filesMappingFile.getName());
        **if (!filesMappingFile.exists())** {
            Log.e("termux", "No file mapping at " + filesMappingFile.getAbsolutePath());
            return;
        }

I keep digging on what's going on because it only fails to execute if I use the following Run option(APK from app Bundle):
![image](https://user-images.githubusercontent.com/11146636/144635426-eea1927a-8962-425a-b513-96707f728e72.png)

when I run it using the Default APK the symlinks is succesfully located and packages are installed succesfully. Do you recall any code section in which a different run could make a difference i the code flow?

**Reply by agnostic-apollo** [reply]
Maybe native libs are not being added to the app bundle. I can look into it later and report back. You can also investigate yourself by opening bundle in android studio by dragging file into it. I haven't played around much with bundles so don't know how android studio handles them. Maybe try adding `android.bundle.enableUncompressedNativeLibs=false` to root level `gradle.properties`.

**Reply by simbadMarino** [reply]
Thanks a lot man!, looking into it right now, will report back as well if I find something

**Reply by agnostic-apollo** [reply]
You are welcome!

**Reply by simbadMarino** [reply]
@agnostic-apollo writing a`ndroid.bundle.enableUncompressedNativeLibs=false` to root level `gradle.properties` fixed the problem! 

I'll make a final test uploading the bundle to Google Play Console and download it from there as internal testing but looks like finally fixed :) Thanks a lot for your support!

BTW: I have a nice proposal to fund termux project and give back a little bit wth the product I'm building :), will provide more details in a different conversation

**Reply by agnostic-apollo** [reply]
> fixed the problem!

Great! Thanks for reporting and helping in fixing in yet another problem :)

>I have a nice proposal to fund termux project 

Cool. Thanks in advance!

**Reply by simbadMarino** [reply]
Change Committed in: https://github.com/simbadMarino/dCloud/commit/b117322887858c528906318d5e7475564ebc9820. 
Just in case anyone is willing to copy/analyze the solution ;)

**Reply by agnostic-apollo** [reply]
@simbadMarino You should fix the vulnerability.

https://github.com/simbadMarino/dCloud/blob/v1.3.2_dCloud/android/app/src/main/AndroidManifest.xml#L193

https://termux.github.io/general/2022/02/15/termux-apps-vulnerability-disclosures.html#termux-files-world-readable

**Reply by simbadMarino** [reply]
Hey @agnostic-apollo , thanks for the notification!! For sure will take a look at it and fix it asap

**Reply by agnostic-apollo** [reply]
You are welcome.

**Reply by simbadMarino** [reply]
Fixed on https://github.com/simbadMarino/dCloud/commit/9af654871e4a9697a3acdf7b7b052de34880975e . Thanks again mate!

**Reply by agnostic-apollo** [reply]
You are very welcome.

But why are you running your app as `sharedUserId` and custom permissions of `com.termux`?

Can't change it now without uninstall.

https://github.com/simbadMarino/dCloud/blob/9af654871e4a9697a3acdf7b7b052de34880975e/android/app/src/main/AndroidManifest.xml#L6

https://github.com/simbadMarino/dCloud/blob/9af654871e4a9697a3acdf7b7b052de34880975e/android/app/build.gradle#L171

https://github.com/simbadMarino/dCloud/blob/9af654871e4a9697a3acdf7b7b052de34880975e/android/termux-shared/src/main/java/com/termux/shared/termux/TermuxConstants.java#L217

Your enfored `targetSdkVersion` will also depend on if termux app was install by user.

https://cs.android.com/android/platform/superproject/+/android-12.0.0_r32:frameworks/base/services/core/java/com/android/server/pm/SharedUserSetting.java;l=195

https://cs.android.com/android/platform/superproject/+/android-12.0.0_r32:frameworks/base/services/core/java/com/android/server/pm/SELinuxMMAC.java;l=352

**Reply by simbadMarino** [reply]
Sorry, I'm trying to understand what you are pointing at... the sharedUserId for sure I got it from the updated-flavour , maybe i didnt understand it's purpose and need to remove it?
About the customer permissions of com.termux what do you mean? As far as I remember I had to keep the com.termux package name as I'm still keeping most of the termux code intact.

If I understood from the definition of what a sharedUserId is I will have installing conflicts, etc  with the official termux app (I actually noticed this problem in the past but pushed it for future analysis). Do you sugget to just remove this line in the manifest and to have a default one generated by the Android system?

**Reply by agnostic-apollo** [reply]
By using the same `sharedUserId`, your app and termux apps are granted the same user id by android and can directly access each others data. Android would not allow installation if there is signature difference between your APK and termux APKs. At this point, you can't just remove the line, since that what imply your app should be assigned a different user id, which won't be possible unless all your users uninstall your app and reinstall the app with an APK that does not have `com.termux` `sharedUserId`. It [may be possible in android 13](https://commonsware.com/blog/2022/02/12/random-musings-android-13-dp1.html#what-else-might-break-your-apps-in-the-not-too-distant-future), but not on earlier versions.

The docs were clear that you must change `TERMUX_PACKAGE_NAME` manifest placeholder.

https://github.com/simbadMarino/dCloud/blob/9af654871e4a9697a3acdf7b7b052de34880975e/android/termux-shared/src/main/java/com/termux/shared/termux/TermuxConstants.java#L157

>About the customer permissions of com.termux what do you mean? As far as I remember I had to keep the com.termux package name as I'm still keeping most of the termux code intact.

Your app should have its own custom permissions (`com.<package_name>.RUN_COMMAND`) declared. Android does not allow two apps with the same permission declared to be installed unless they share the same `sharedUserId` due to security reasons.

As for `package="com.termux"` in `AndroidManifest.xml`, its mainly used at compile time for `.TermuxActivity` expansion. You have already changed that and are using full `com.termux.app.TermuxActivity` class declarations, so that's not an issue. Once APK is build the `package` value is replaced by `applicationId` in the `AndroidManifest.xml` of the APK.

I suggest you read https://developer.android.com/studio/build/configure-app-module

You probably won't be allowed to publish on playstore either due to conflicts, since I think only fornwall's account should be allowed to publish app with that user id, at least that would be a good security practice.

**Reply by simbadMarino** [reply]
Thanks again for your feedback @agnostic-apollo , so the last 2 days I've been trying to get the `sharedUserId` problem fixed without success... I followed the TermuxConstant guidelines thoughtfully and changed all hardcoded `com.termux` every time I try to change the the `sharedUserId` I get the following error:


W/dCloud:SharedProperties: Not loading properties since file is null
E/FMQ: grantorIdx must be less than 3
D/CCodecBufferChannel: [c2.android.vorbis.decoder#733] MediaCodec discarded an unknown buffer
I/chatty: uid=10173(com.justshare) NDK MediaCodec_ identical 1 line
D/CCodecBufferChannel: [c2.android.vorbis.decoder#733] MediaCodec discarded an unknown buffer
I/hw-BpHwBinder: onLastStrongRef automatically unlinking death recipients
D/BufferPoolAccessor2.0: bufferpool2 0x7d9d56a268 : 0(0 size) total buffers - 0(0 size) used buffers - 3/9 (recycle/alloc) - 6/24 (fetch/transfer)
D/BufferPoolAccessor2.0: evictor expired: 1, evicted: 1


As soon as I change the `TERMUX_PACKAGE_NAME` manifest placeholder terminal is not properly loaded and the error above pops up.

Is there something else I need to change in the `com.termux.shared` I need to modify maybe?

**Reply by agnostic-apollo** [reply]
Another bug, `clean` gradle task does not clear `app/src/main/jniLibs/`, so any package files that are not overridden may remain there and consume space in apk, although shouldn't be extracted to prefix because of not existing in `libfiles.so`. Requires a patch like following, assuming there are no other libs in `jniLibs`. Also not cleaning zips, since one may have custom ones there.

https://github.com/termux/termux-app/blob/updated-flavor/app/build.gradle

```
clean {
    doLast {
        def tree = fileTree(new File(projectDir, 'src/main'))
        //tree.include 'bootstraps/bootstrap-*.zip'
        tree.include 'bootstraps/bootstrap-*.zip.*.done'
        tree.include 'jniLibs/*/*'
        tree.each { it.delete() }
    }
}
```

**Reply by simbadMarino** [reply]
Nice, I noticed it wasn’t removing those and sometimes I used to clean them up manually (specially when releasing a new apk) I will include this change in the next iteration! Thank you!
